lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -113,15 +113,36 @@ defmodule KinesisClient.Stream.Shard.Producer do
     {:noreply, %{state | shard_closed_timer: timer}}
   end
 
+  # TODO(KNO-2531) Switch to a custom Broadway.Acknowledger instead of relying
+  # on Broadway.CallerAcknowledger so that we can more efficiently checkpoint
+  # our progress in dynamo.
   @impl GenStage
-  def handle_info({:ack, _ref, successful_msgs, []}, state) do
-    %{metadata: %{"SequenceNumber" => checkpoint}} = successful_msgs |> Enum.reverse() |> hd()
+  def handle_info({:ack, _ref, successful_msgs, failed_msgs}, state) do
+    checkpoint = get_top_checkpoint(successful_msgs, failed_msgs)
+
+    # -1 means there wasn't a reliable checkpoint figure present, so
+    # we should not update the checkpoint at all.
+    # This should never happen with conforming messages.
+    if checkpoint != "-1" do
+      :ok = update_checkpoint(state, checkpoint)
+    else
+      Logger.warn("""
+        Unable to update checkpoint for app_name: #{state.app_name}, shard_id: #{state.shard_id}
+
+        This might happen for a few reasons (in order of likelihood):
+        1. You are writing tests, and your messages need something like %{metadata: %{"SequenceNumber" => "SomeStringHere"}}
+        2. Kinesis changed how it publishes messages
+        3. We tried to ack an empty batch (Not possible. Then again, we also don't _really_ know why the sky is blue, so go check your confidence.)
+      """)
```

Comment:
Idea: here (and the new raises in `update_checkpoint`), it might be useful to emit some more custom Telemetry events so that we can instrument this edges in our dashboards. Basically expanding upon this work: https://github.com/knocklabs/kcl_ex/pull/5

Thoughts?

Reply:
This is a really fringe tight edge case, so we'd need to add an alarm for this particular case. I wrote this more for satisfying testing than anything, and to provide some clear directions _if_ we hit this code path. But, I honestly never expect to see this message in production, since messages should always conform to the spec from AWS, and we shouldn't get this far without having a batch to work with.

Reply:
For the other raises, I can wrap those in custom structs, although I think the fact that we are wrapped in a `span` will capture the exception, right? These are exceptions in the most literal sense.

Reply:
> But, I honestly never expect to see this message in production, since messages should always conform to the spec from AWS, and we shouldn't get this far without having a batch to work with.

Cool, makes sense 👍 

> although I think the fact that we are wrapped in a span will capture the exception, right

Oh yes absolutely! My suggestion to use the exception structs is purely for visibility in Sentry reports. Personally I find named exceptions a little easier to mentally map then RuntimeErrors with string messages. But that's partially preference and I don't think this is blocking at all.